src/python/pants/backend/python/typecheck/mypy/rules.py

```diff
@@ -142,59 +182,48 @@ async def mypy_typecheck(
-    #     they have incompatible wheels.
-    #
-    #     Instead, we teach MyPy about the requirements by extracting the distributions from
-    #     requirements.pex and setting EXTRACTED_WHEELS, which our custom launcher script then
-    #     looks for.
-    code_interpreter_constraints = PexInterpreterConstraints.create_from_compatibility_fields(
-        (
-            tgt[PythonInterpreterCompatibility]
-            for tgt in typechecked_transitive_targets.closure
-            if tgt.has_field(PythonInterpreterCompatibility)
-        ),
-        python_setup,
+    # If the user did not set `--python-version` already, we set it ourselves based on their code's
+    # interpreter constraints. This determines what AST is used by MyPy.
+    python_version = (
+        None
+        if partition.python_version_already_configured
+        else partition.interpreter_constraints.minimum_python_version()
     )
 
-    if not mypy.options.is_default("interpreter_constraints"):
-        tool_interpreter_constraints = mypy.interpreter_constraints
-    elif code_interpreter_constraints.requires_python38_or_newer():
-        tool_interpreter_constraints = ("CPython>=3.8",)
-    elif code_interpreter_constraints.requires_python37_or_newer():
-        tool_interpreter_constraints = ("CPython>=3.7",)
-    elif code_interpreter_constraints.requires_python36_or_newer():
-        tool_interpreter_constraints = ("CPython>=3.6",)
-    else:
-        tool_interpreter_constraints = mypy.interpreter_constraints
+    # MyPy requires 3.5+ to run, but uses the typed-ast library to work with 2.7, 3.4, 3.5, 3.6,
+    # and 3.7. However, typed-ast does not understand 3.8, so instead we must run MyPy with
+    # Python 3.8 when relevant. We only do this if if <3.8 can't be used, as we don't want a
+    # loose requirement like `>=3.6` to result in requiring Python 3.8, which would error if
+    # 3.8 is not installed on the machine.
+    tool_interpreter_constraints = PexInterpreterConstraints(
+        ("CPython>=3.8",)
+        if (
+            mypy.options.is_default("interpreter_constraints")
+            and partition.interpreter_constraints.requires_python38_or_newer()
+        )
+        else mypy.interpreter_constraints
+    )
 
     plugin_sources_request = Get(
         PythonSourceFiles, PythonSourceFilesRequest(plugin_transitive_targets.closure)
     )
     typechecked_sources_request = Get(
-        PythonSourceFiles, PythonSourceFilesRequest(typechecked_transitive_targets.closure)
+        PythonSourceFiles, PythonSourceFilesRequest(partition.closure)
     )
 
+    # Normally, this `requirements.pex` would be merged with mypy.pex via `--pex-path`. However,
+    # this will cause a runtime error if the interpreter constraints are different between the
+    # PEXes and they have incompatible wheels.
+    #
+    # Instead, we teach MyPy about the requirements by extracting the distributions from
```

Comment:
It might be good to reference the existence of the MYPYPATH/mypy_path option (presumably we'd use it if we didn't already have to extract in the wrapper?)